### ISSUE TITLE
twisted.mail.relaymanager.MXCalculator._cbMX raises unhandled IndexError for response with no answers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2807]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2807|
|Type|defect|
|Created|2007-08-27 21:08:13Z|

If a hostname has no MX records, ```getMX``` will return a ```Deferred``` with an empty answers list.  ```_cbMX``` doesn't handle this case and raises an IndexError when it tries to get the 0th element at the end of the method.  ```_ebMX``` doesn't handle this exception, so it falls through to user code, which isn't expecting it (it's not documented).

According to [RFC 974](http://www.faqs.org/rfcs/rfc974.html), in the section headed _Interpreting the List of MX RRs_:

```
   It is possible that the list of MXs in the response to the query will
   be empty.  This is a special case.  If the list is empty, mailers
   should treat it as if it contained one RR, an MX RR with a preference
   value of 0, and a host name of REMOTE.  (I.e., REMOTE is its only
   MX).  In addition, the mailer should do no further processing on the
   list, but should attempt to deliver the message to REMOTE.
```

<details><summary>Searchable metadata</summary>

```
trac-id__2807 2807
type__defect defect
reporter__exarkun exarkun
priority__highest highest
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__names names
keywords__ 
time__1188248893000000 1188248893000000
changetime__1189609499000000 1189609499000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun removed owner|
|-|-|

Done in mxcalc-indexerror-2807

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve set owner to @exarkun|
|-|-|

test_successWithoutResults is not totally clear (at least to me), because it uses the successful path of ```_ebMX```. Maybe another test with ```fallbackToDomain``` set to false would clarify this. Also, please remove trailing whitespaces of relaymanager.

That's it!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @therve|
|-|-|

Can you clarify this a bit?

test_sucessWithoutResults definitely only exercises one codepath through _ebMX.  If one sets fallbackToDomain to False and tries this, one will get an IOError.  Is this the case you'd like to see a test for?  Or do you think that even with fallbackToDomain set to false, getMX should be returning a Record_MX with the address from the A record of the hostname?

Either could seem reasonable to me, I just want to know which you're asking for.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @exarkun|
|-|-|

It was the former, but it's not totally justified: my problem was that the diff showed a new failure, whereas your test was a success test, but looking at the code more it totally makes sense, so if you feel like adding a new test, great, otherwise please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Okay, great.  Since I didn't really touch any code in _ebMX, I don't feel a strong urge to add another test covering that case, so I think I'll just merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [21279]) Merge mxcalc-indexerror-2807

Author: exarkun
Reviewer: therve
Fixes [#2807](https://github.com/twisted/twisted/issues/2807)

In MXCalculator, handle the case where an MX lookup succeeds but there are no
answers.  The defines behavior for this case, now MXCalculator properly
implements it.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
